test-suite/src/test/scala/org/scalajs/testsuite/compiler/OptimizerTest.scala

```diff
@@ -81,6 +81,64 @@ object OptimizerTest extends JasmineTest {
 
   }
 
+  describe("String_+ constant folding") {
+    it("must fold two constant string to the correct value") {
+      expect("I am " + "constant").toEqual("I am constant")
+    }
+
+    it("must fold the empty string when associated with a string") {
+      val number: String = "1" + (1.4f).toString.substring(1)
+      expect(number + "").toEqual("1.399999976158142")
+      expect("" + number).toEqual("1.399999976158142")
+    }
+
+    it("must fold 1.4f and a stringLit to the correct value") {
+      expect(1.4f + "hello").toEqual("1.399999976158142hello")
+      expect("hello" + 1.4f).toEqual("hello1.399999976158142")
+    }
+
+    it("must fold cascading String + correctly") {
+      def someString(a: Int) = "awesome! " + a + "/10"
+      expect("ScalaJS" + (" is " + someString(10))).toEqual("ScalaJS is awesome! 10/10")
+      expect((someString(10) + " is ") + "ScalaJS").toEqual("awesome! 10/10 is ScalaJS")
+    }
+
+    it("must fold integer in double and stringLit correctly") {
+      expect(1.0 + "hello").toEqual("1hello")
+      expect("hello" + 1.0).toEqual("hello1")
+    }
+
+    it("must fold double with decimal and stringLit correctly") {
+      expect(1.1 + "hello").toEqual("1.1hello")
+      expect("hello" + 1.1).toEqual("hello1.1")
+    }
+
+    it("must fold long and stringLit correctly") {
+      expect(1L + "hello").toEqual("1hello")
+      expect("hello" + 1L).toEqual("hello1")
+    }
+
+    it("must fold integer and stringLit correctly") {
+      expect(42 + "hello").toEqual("42hello")
+      expect("hello" + 42).toEqual("hello42")
+    }
+
+    it("must fold boolean and stringLit correctly") {
+      expect(false + " is not true").toEqual("false is not true")
+      expect("false is not " + true).toEqual("false is not true")
+    }
+
+    it("must fold unit and stringLit correctly") {
+      expect(() + " is undefined?").toEqual("undefined is undefined?")
+      expect("undefined is " + ()).toEqual("undefined is undefined")
+    }
+
+    it("must fold null and stringLit correctly") {
+      expect("Damien is not " + null).toEqual("Damien is not null")
+    }
+
+  }
```

Comment:
It would be good to also have a test with `Char`. Not sure it will actually constant-fold, but it would be good to guard against a mistake, somewhere.